fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java

```diff
@@ -357,18 +358,30 @@ public void sendVote(List<ElectionVote> votes) {
 
     try {
       PoPToken token = keyManager.getValidPoPToken(lao);
-      CastVote castVote = new CastVote(votes, election.getId(), lao.getId());
-      Channel electionChannel = election.getChannel();
+      CastVote<ElectionVote> castOpenBallotVote = null;
+      CastVote<ElectionEncryptedVote> castEncryptedVote = null;
+
+      // Construct the cast vote depending if the messages need to be encrypted or not
+      if (election.getElectionVersion() == ElectionVersion.OPEN_BALLOT) {
+        castOpenBallotVote = new CastVote<>(votes, election.getId(), lao.getId());
+      } else {
+        List<ElectionEncryptedVote> encryptedVotes = election.encrypt(votes);
+        castEncryptedVote = new CastVote<>(encryptedVotes, election.getId(), lao.getId());
+        Toast.makeText(getApplication(), "Vote encrypted !", Toast.LENGTH_LONG)
+                .show();
+      }
 
+      Channel electionChannel = election.getChannel();
       Log.d(TAG, PUBLISH_MESSAGE);
       Disposable disposable =
-          networkManager
-              .getMessageSender()
-              .publish(token, electionChannel, castVote)
-              .doFinally(this::openLaoDetail)
-              .subscribe(
-                  () -> {
-                    Log.d(TAG, "sent a vote successfully");
+              networkManager
+                      .getMessageSender()
+                      .publish(token, electionChannel,
+                              election.getElectionVersion() == ElectionVersion.OPEN_BALLOT ? castOpenBallotVote : castEncryptedVote)
+                      .doFinally(this::openLaoDetail)
+                      .subscribe(
+                              () -> {
+                                Log.d(TAG, "sent a vote successfully");
```

Comment:
This is weird.

The control flow is hard to understand.
Could you try to create sub-functions such that the code is easier to read ?